### DISCUSSION TITLE
I really don't know how to solve this problem

### DISCUSSION THREAD

**Discussion by NOVA-ME** [tokio-rs/axum#263] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/18442141/130632300-a41d0c2b-32fd-4790-9071-90112f71bcf0.png)
![image](https://user-images.githubusercontent.com/18442141/130632320-7acbbc27-2d8b-4c16-a27f-ff289a91147b.png)
![image](https://user-images.githubusercontent.com/18442141/130632400-507a2d79-f227-4be7-a150-bacececcb13f.png)
![image](https://user-images.githubusercontent.com/18442141/130632436-53395334-5dc0-4d73-9027-fda87b62130f.png)
only the ```get_captcha``` method can't add to router, ervery thing else is fine.
And I want to ask there have the axum telegram group or other place, that I can ask for some question

**Comment by davidpdrsn** [comment, marked as answer]
Try to isolate which part of the function makes it not work

- Replace the body of the function with `todo!()`.
- If that works then try and remove each of the lines until you find which line might be causing it. Note that handlers must return futures that are `Send`, but that doesn't seem to be an issue here.
- If it doesn't work then try changing the return type to simply `()` or `(StatusCode, HeaderMap, Vec<u8>)`. Maybe `AppResult`'s error type doesn't implement `IntoResponse`?

**Reply by NOVA-ME** [reply]
![image](https://user-images.githubusercontent.com/18442141/130639541-4ed536d3-2631-42b9-b6ba-25a9f245f0a4.png)
Thank you for your reply. I learn a lot from this.
I think this is the reason. But I don't know why, I need return a image captcha
[captcha crate](https://crates.io/crates/captcha)
I finally solve it, I move the ```redis_set``` out of the function.

**Reply by davidpdrsn** [reply]
I don't quite understand 😅 Did you fix it or not?

If not, could you make a minimal reproduction script that I can download the play with myself?

**Reply by NOVA-ME** [reply]
> I don't quite understand  Did you fix it or not?
> 
> If not, could you make a minimal reproduction script that I can download the play with myself?

I have soled it. Move the ```redis_get``` out of the method, it works fine, but I don't know why, I'm noob in rust async/await.

**Reply by davidpdrsn** [reply]
Which library does `redis_get` come from?

**Reply by NOVA-ME** [reply]
> Which library does `redis_get` come from?
![image](https://user-images.githubusercontent.com/18442141/130644499-1910bf65-1b1a-41a2-8f7e-9d91d6249dd6.png)
[redis crate](https://crates.io/crates/redis)
[r2d2 crate](https://crates.io/crates/r2d2)

**Reply by davidpdrsn** [reply]
Hm alright. I'm not familiar enough with those crates to know. It would seem somewhere a `!Send` future shows up, which axum doesn't support.

**Reply by NOVA-ME** [reply]
> > Hm alright. I'm not familiar enough with those crates to know. It would seem somewhere a `!Send` future shows up, which axum doesn't support.
> 
> Thank you so much. The axum 0.2 is quite nice.

**Reply by mineralres** [reply]
@davidpdrsn @jkristell  I got the similar problem,  please check my code if you have time, thanks.

https://github.com/tokio-rs/axum/issues/299

**Reply by davidpdrsn** [reply]
@mineralres You don't need to ping me twice.

**Comment by jkristell** [comment]
Had a similar problem. The problem was the RwLockWriteGuard from std::sync (it is !Send). Switching to RwLock from tokio solved my problem.
